contracts/random/RandomSeedProvider.sol

```diff
@@ -0,0 +1,286 @@
+        if (randomSource == ONCHAIN || !approvedForOffchainRandom[msg.sender]) {
+            // Generate a value for this block if one has not been generated yet. This
+            // is required because there may have been calls to requestRandomSeed
+            // in previous blocks that are waiting for a random number to be produced.
+            _generateNextRandomOnChain();
+
+            // Indicate that a value based on the next block will be fine.
+            _randomFulfillmentIndex = nextRandomIndex;
+
+            _randomSource = ONCHAIN;
+        } else {
+            // Limit how often offchain random numbers are requested to a maximum of once per block.
+            if (lastBlockOffchainRequest == block.number) {
+                _randomFulfillmentIndex = prevOffchainRandomRequest;
+            } else {
+                _randomFulfillmentIndex = IOffchainRandomSource(randomSource).requestOffchainRandom();
+                prevOffchainRandomRequest = _randomFulfillmentIndex;
+                lastBlockOffchainRequest = block.number;
+            }
+            _randomSource = randomSource;
+        }
+    }
+
+    /**
+     * @notice Fetches a random seed value that was requested using requestRandom.
+     * @dev Note that the same _randomSeed will be returned to multiple games and even within
+     * @dev the one game. Games must personalise this value to their own game, the the particular game player,
+     * @dev and to the game player's request.
+     * @return _randomSeed The value from with random values can be derived.
+     */
+    function getRandomSeed(
+        uint256 _randomFulfillmentIndex,
+        address _randomSource
+    ) external returns (bytes32 _randomSeed) {
+        if (_randomSource == ONCHAIN) {
+            _generateNextRandomOnChain();
+            if (_randomFulfillmentIndex >= nextRandomIndex) {
+                revert WaitForRandom();
+            }
+            return randomOutput[_randomFulfillmentIndex];
+        } else {
+            // If random source is not the address of a valid contract this will revert
+            // with no revert information returned.
+            return IOffchainRandomSource(_randomSource).getOffchainRandom(_randomFulfillmentIndex);
+        }
+    }
+
+    /**
+     * @notice Check whether a random seed is ready.
+     * @param _randomFulfillmentIndex Index when random seed will be ready.
+     */
+    function isRandomSeedReady(uint256 _randomFulfillmentIndex, address _randomSource) external view returns (bool) {
+        if (_randomSource == ONCHAIN) {
+            if (lastBlockRandomGenerated == block.number) {
+                return _randomFulfillmentIndex < nextRandomIndex;
+            } else {
+                return _randomFulfillmentIndex < nextRandomIndex + 1;
+            }
+        } else {
+            return IOffchainRandomSource(randomSource).isOffchainRandomReady(_randomFulfillmentIndex);
```

Comment:
should this be using `_randomSource` instead of `randomSource` ?

Reply:
Well spotted. This is a bug: the random source to use should be the one passed in. I have a test for switching offchain sources. However, the test does not yet check that isRandomReady returns the correct value. I will update the test, see it failing, and then fix the code.